pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```
@@ -24,307 +24,480 @@
+  // For LONG_MAP_BASED storage type.
   private Long2IntOpenHashMap _groupKeyToId;
+  // For ARRAY_MAP_BASED storage type.
+  private Object2IntOpenHashMap<IntArrayList> _arrayGroupKeyToId;
 
   /**
    * Constructor for the class. Initializes data members (reusable arrays).
    *
-   * @param dataFetcher
-   * @param groupByColumns
-   * @param maxNumGroupKeys
+   * @param dataFetcher data fetcher.
+   * @param groupByColumns group-by columns.
    */
-  DefaultGroupKeyGenerator(DataFetcher dataFetcher, String[] groupByColumns, int maxNumGroupKeys) {
+  DefaultGroupKeyGenerator(DataFetcher dataFetcher, String[] groupByColumns) {
     _numGroupByColumns = groupByColumns.length;
+    _cardinalities = new int[_numGroupByColumns];
+    _isSingleValueGroupByColumn = new boolean[_numGroupByColumns];
     _dictionaries = new Dictionary[_numGroupByColumns];
     _singleBlockValSets = new BlockValSet[_numGroupByColumns];
     _multiValIterators = new BlockMultiValIterator[_numGroupByColumns];
-    _cardinalities = new int[_numGroupByColumns];
-    _isSingleValueGroupByColumn = new boolean[_numGroupByColumns];
     _reusableSingleDictIds = new int[_numGroupByColumns][];
 
+    // Track the max number of values among all multi value group-by columns.
     int maxNumMultiValues = 0;
+
+    boolean longOverflow = false;
     for (int i = 0; i < _numGroupByColumns; i++) {
       DataSource dataSource = dataFetcher.getDataSourceForColumn(groupByColumns[i]);
-      _dictionaries[i] = dataSource.getDictionary();
-      _cardinalities[i] = dataSource.getDataSourceMetadata().cardinality();
-      _isSingleValueGroupByColumn[i] = dataSource.getDataSourceMetadata().isSingleValue();
 
+      // Store group-by column cardinalities and update cardinality product.
+      int cardinality = dataSource.getDataSourceMetadata().cardinality();
+      _cardinalities[i] = cardinality;
+      if (!longOverflow) {
+        if (_cardinalityProduct > Long.MAX_VALUE / cardinality) {
```

I cannot think of this kind of cases. 0 cardinality means no value for this column (even no default value)? In that case maybe just throw an divideByZero exception is fine?